kernel_tuner/strategies/diff_evo.py

```diff
@@ -1,42 +1,410 @@
-                                        polish=False, strategy=method, disp=tuning_options.verbose, x0=x0)
+        differential_evolution(searchspace, cost_func, bounds, popsize, maxiter, F, CR, method, constraint_aware, tuning_options.verbose)
     except util.StopCriterionReached as e:
         if tuning_options.verbose:
             print(e)
 
-    if opt_result and tuning_options.verbose:
-        print(opt_result.message)
-
     return cost_func.results
 
 
 tune.__doc__ = common.get_strategy_docstring("Differential Evolution", _options)
+
+
+def values_to_indices(individual_values, tune_params):
+    """Converts an individual's values to its corresponding index vector."""
+    idx = np.zeros(len(individual_values))
+    for i, v in enumerate(tune_params.values()):
+        idx[i] = v.index(individual_values[i])
+    return idx
+
+
+def indices_to_values(individual_indices, tune_params):
+    """Converts an individual's index vector back to its values."""
+    tune_params_list = list(tune_params.values())
+    values = []
+    for dim, idx in enumerate(individual_indices):
+        values.append(tune_params_list[dim][idx])
+    return values
+
+
+def parse_method(method):
+    """Helper func to parse the preferred method into its components."""
+    pattern = r"^(best|rand|currenttobest|randtobest)(1|2)(bin|exp)$"
+    match = re.fullmatch(pattern, method)
+
+    if match:
+        if match.group(1) in ["currenttobest", "randtobest"]:
+            mutation_method = mutation[match.group(1)]
+        else:
+            mutation_method = mutation[match.group(2)]
+        return match.group(1) == "best", int(match.group(2)), mutation_method, crossover[match.group(3)]
+    else:
+        raise ValueError("Error parsing differential evolution method")
+
+
+def random_draw(idxs, mutate, best):
+    """
+    Draw requested number of random individuals.
+
+    Draw without replacement unless there is not enough to draw from.
+    """
+    draw = 2 * mutate + 1 - int(best)
+    return np.random.choice(idxs, draw, replace=draw >= len(idxs))
+
+
+def generate_population(tune_params, max_idx, popsize, searchspace, constraint_aware):
+    """ Generate new population, returns Numpy array """
+    if constraint_aware:
+        samples = LatinHypercube(len(tune_params)).integers(l_bounds=0, u_bounds=max_idx, n=popsize, endpoint=True)
+        population = [indices_to_values(sample, tune_params) for sample in samples]
+        population = [repair(individual, searchspace) for individual in population]
+    else:
+        population = []
+        for _ in range(popsize):
+            ind = []
+            for key in tune_params:
+                ind.append(random.choice(tune_params[key]))
+            population.append(ind)
+    return population
+
+
+def differential_evolution(searchspace, cost_func, bounds, popsize, maxiter, F, CR, method, constraint_aware, verbose):
```

Comment:
Officieel moeten in Python variabele namen voldoen aan deze regex: `[^[_a-z][a-z0-9_]*$.]`